From 4fc863f6c8a3752d10542707162b3bd57d55e09a Mon Sep 17 00:00:00 2001
From: Liang Qi <liang.qi@qt.io>
Date: Wed, 26 May 2021 10:32:04 +0200
Subject: [PATCH] xcb: fix QWindow::startSystemMove()/Resize() triggered by
 mouse
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We can't get mouse release event from master pointers after
QXcbWindow::doStartSystemMoveResize() which calls xcb_ungrab_pointer(),
it looks like most X11 WMs work as that.

So we try to get mouse release event from slave pointers.

Based on https://specifications.freedesktop.org/wm-spec/1.4/ar01s04.html
, we need to send _NET_WM_MOVERESIZE_CANCEL when we get mouse release
event.

Task-number: QTBUG-91077
Change-Id: I01e74a01c87b381ee7cd6f20d51a1fa61c0e98fc
Reviewed-by: Tor Arne Vestbø <tor.arne.vestbo@qt.io>
(cherry picked from commit 8e506fdd299d2fa18172209bea316f484e234e19)
(cherry picked from commit 03def168eac535541a8ae6f48c043624c90bb58a)
---
 src/plugins/platforms/xcb-maemo/qxcbconnection.cpp  |  3 ++
 src/plugins/platforms/xcb-maemo/qxcbconnection.h    | 10 +++-
 .../platforms/xcb/qxcbconnection_xi2.cpp      | 50 +++++++++++++++++--
 src/plugins/platforms/xcb-maemo/qxcbwindow.cpp      |  8 +--
 4 files changed, 62 insertions(+), 9 deletions(-)

Index: qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection.cpp
===================================================================
--- qt-platform-maemo.orig/src/plugins/platforms/xcb-maemo/qxcbconnection.cpp
+++ qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection.cpp
@@ -90,6 +90,7 @@ Q_LOGGING_CATEGORY(lcQpaXDnd, "qt.qpa.xd
 
 QXcbConnection::QXcbConnection(QXcbNativeInterface *nativeInterface, bool canGrabServer, xcb_visualid_t defaultVisualId, const char *displayName)
     : QXcbBasicConnection(displayName)
+    , m_duringSystemMoveResize(false)
     , m_canGrabServer(canGrabServer)
     , m_defaultVisualId(defaultVisualId)
     , m_nativeInterface(nativeInterface)
@@ -604,6 +605,8 @@ void QXcbConnection::handleXcbEvent(xcb_
     }
     case XCB_BUTTON_RELEASE: {
         auto ev = reinterpret_cast<xcb_button_release_event_t *>(event);
+        if (m_duringSystemMoveResize && ev->root != XCB_NONE)
+            abortSystemMoveResize(ev->root);
         m_keyboard->updateXKBStateFromCore(ev->state);
         m_buttonState = (m_buttonState & ~0x7) | translateMouseButtons(ev->state);
         setButtonState(translateMouseButton(ev->detail), false);
Index: qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection.h
===================================================================
--- qt-platform-maemo.orig/src/plugins/platforms/xcb-maemo/qxcbconnection.h
+++ qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection.h
@@ -230,10 +230,13 @@ public:
     bool xi2MouseEventsDisabled() const;
     Qt::MouseButton xiToQtMouseButton(uint32_t b);
     void xi2UpdateScrollingDevices();
-    bool startSystemMoveResizeForTouch(xcb_window_t window, int edges);
-    void abortSystemMoveResizeForTouch();
     bool isTouchScreen(int id);
 
+    bool startSystemMoveResizeForTouch(xcb_window_t window, int edges);
+    void abortSystemMoveResize(xcb_window_t window);
+    bool isDuringSystemMoveResize() const;
+    void setDuringSystemMoveResize(bool during);
+
     bool canGrab() const { return m_canGrabServer; }
 
     QXcbGlIntegration *glIntegration() const;
@@ -330,12 +333,14 @@ private:
     static bool xi2GetValuatorValueIfSet(const void *event, int valuatorNum, double *value);
 
     QHash<int, TouchDeviceData> m_touchDevices;
+
     struct StartSystemMoveResizeInfo {
         xcb_window_t window = XCB_NONE;
         uint16_t deviceid;
         uint32_t pointid;
         int edges;
     } m_startSystemMoveResizeInfo;
+    bool m_duringSystemMoveResize;
 
     const bool m_canGrabServer;
     const xcb_visualid_t m_defaultVisualId;
@@ -374,6 +379,7 @@ private:
     mutable bool m_glIntegrationInitialized = false;
     bool m_xiGrab = false;
     QVector<int> m_xiMasterPointerIds;
+    QVector<int> m_xiSlavePointerIds;
 
     xcb_window_t m_qtSelectionOwner = 0;
 
Index: qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection_xi2.cpp
===================================================================
--- qt-platform-maemo.orig/src/plugins/platforms/xcb-maemo/qxcbconnection_xi2.cpp
+++ qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbconnection_xi2.cpp
@@ -89,7 +89,7 @@ void QXcbConnection::xi2SelectDeviceEven
     }
 
     qt_xcb_input_event_mask_t mask;
-    mask.header.deviceid = XCB_INPUT_DEVICE_ALL_MASTER;
+    mask.header.deviceid = XCB_INPUT_DEVICE_ALL;
     mask.header.mask_len = 1;
     mask.mask = bitMask;
     xcb_void_cookie_t cookie =
@@ -304,8 +304,10 @@ void QXcbConnection::xi2SetupDevices()
             continue;
         }
         // only slave pointer devices are relevant here
-        if (deviceInfo->type == XCB_INPUT_DEVICE_TYPE_SLAVE_POINTER)
+        if (deviceInfo->type == XCB_INPUT_DEVICE_TYPE_SLAVE_POINTER) {
+            m_xiSlavePointerIds.append(deviceInfo->deviceid);
             xi2SetupDevice(deviceInfo, false);
+        }
     }
 
     if (m_xiMasterPointerIds.size() > 1)
@@ -523,6 +525,19 @@ static inline qreal fixed1616ToReal(xcb_
 void QXcbConnection::xi2HandleEvent(xcb_ge_event_t *event)
 {
     auto *xiEvent = reinterpret_cast<qt_xcb_input_device_event_t *>(event);
+    if (m_xiSlavePointerIds.contains(xiEvent->deviceid)) {
+        if (!m_duringSystemMoveResize)
+            return;
+        if (xiEvent->event == XCB_NONE)
+            return;
+
+        if (xiEvent->event_type == XCB_INPUT_BUTTON_RELEASE
+            && xiEvent->detail == XCB_BUTTON_INDEX_1 ) {
+            abortSystemMoveResize(xiEvent->event);
+        } else {
+            return;
+        }
+    }
     int sourceDeviceId = xiEvent->deviceid; // may be the master id
     qt_xcb_input_device_event_t *xiDeviceEvent = nullptr;
     xcb_input_enter_event_t *xiEnterEvent = nullptr;
@@ -831,9 +846,38 @@ bool QXcbConnection::startSystemMoveResi
     return false;
 }
 
-void QXcbConnection::abortSystemMoveResizeForTouch()
+void QXcbConnection::abortSystemMoveResize(xcb_window_t window)
 {
+    qCDebug(lcQpaXInputDevices) << "sending client message NET_WM_MOVERESIZE_CANCEL to window: " << window;
     m_startSystemMoveResizeInfo.window = XCB_NONE;
+
+    const xcb_atom_t moveResize = connection()->atom(QXcbAtom::_NET_WM_MOVERESIZE);
+    xcb_client_message_event_t xev;
+    xev.response_type = XCB_CLIENT_MESSAGE;
+    xev.type = moveResize;
+    xev.sequence = 0;
+    xev.window = window;
+    xev.format = 32;
+    xev.data.data32[0] = 0;
+    xev.data.data32[1] = 0;
+    xev.data.data32[2] = 11; // _NET_WM_MOVERESIZE_CANCEL
+    xev.data.data32[3] = 0;
+    xev.data.data32[4] = 0;
+    xcb_send_event(xcb_connection(), false, primaryScreen()->root(),
+                   XCB_EVENT_MASK_SUBSTRUCTURE_REDIRECT | XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY,
+                   (const char *)&xev);
+
+    m_duringSystemMoveResize = false;
+}
+
+bool QXcbConnection::isDuringSystemMoveResize() const
+{
+    return m_duringSystemMoveResize;
+}
+
+void QXcbConnection::setDuringSystemMoveResize(bool during)
+{
+    m_duringSystemMoveResize = during;
 }
 
 bool QXcbConnection::xi2SetMouseGrabEnabled(xcb_window_t w, bool grab)
Index: qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbwindow.cpp
===================================================================
--- qt-platform-maemo.orig/src/plugins/platforms/xcb-maemo/qxcbwindow.cpp
+++ qt-platform-maemo/src/plugins/platforms/xcb-maemo/qxcbwindow.cpp
@@ -2691,14 +2691,11 @@ bool QXcbWindow::startSystemMoveResize(c
     if (startedByTouch) {
         if (connection()->isUnity()) {
             // Unity fails to move/resize via _NET_WM_MOVERESIZE (WM bug?).
-            connection()->abortSystemMoveResizeForTouch();
+            connection()->abortSystemMoveResize(m_window);
             return false;
         }
         // KWin, Openbox, AwesomeWM and Gnome have been tested to work with _NET_WM_MOVERESIZE.
     } else { // Started by mouse press.
-        if (connection()->isUnity())
-            return false; // _NET_WM_MOVERESIZE on this WM is bouncy (WM bug?).
-
         doStartSystemMoveResize(mapToGlobal(pos), edges);
     }
 
@@ -2730,6 +2727,7 @@ static uint qtEdgesToXcbMoveResizeDirect
 
 void QXcbWindow::doStartSystemMoveResize(const QPoint &globalPos, int edges)
 {
+    qCDebug(lcQpaXInputDevices) << "triggered system move or resize via sending _NET_WM_MOVERESIZE client message";
     const xcb_atom_t moveResize = connection()->atom(QXcbAtom::_NET_WM_MOVERESIZE);
     xcb_client_message_event_t xev;
     xev.response_type = XCB_CLIENT_MESSAGE;
@@ -2749,6 +2747,8 @@ void QXcbWindow::doStartSystemMoveResize
     xcb_send_event(connection()->xcb_connection(), false, xcbScreen()->root(),
                    XCB_EVENT_MASK_SUBSTRUCTURE_REDIRECT | XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY,
                    (const char *)&xev);
+
+    connection()->setDuringSystemMoveResize(true);
 }
 
 // Sends an XEmbed message.
